java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java

```diff
@@ -231,11 +224,9 @@ protected RecordReaderImpl(ReaderImpl fileReader,
     if (options.getPreFilterColumnNames() != null) {
       for (String colName : options.getPreFilterColumnNames()) {
         int expandColId = findColumns(evolution, colName);
+        // If the column is not present in the file then this can be ignored from read.
         if (expandColId != -1) {
           filterColIds.add(expandColId);
-        } else {
```

Comment:
The intention of the original code here was to notify the consumer that the filter can not be applied  and could lead to wrong results (keep in mind that row-filtering is not best-effort as row-group skipping and the consumer expects only the rows that actually pass the filter).
This was also a side-effect of the existing code that had no distinction between a column missing from the schema, and schema-fileReader missmatch as you very well mentioned above.

However, (in the former case) we can now end up with pretty generic exceptions from findSubtype that do not reveal the root cause:
**IllegalArgumentException("Field " + first + " not found in " + current.toString())**

I believe the best thing to do here (even if its a bit ugly) is capture the Exception and rethrow with an appropriate message.

Reply:
Agreed. Wrapped the exception in the method as suggested.